### PR TITLE
Changed collection method of the second counter value in the Windows Perfmon module of Metricbeat

### PR DESCRIPTION
**Pull request by jorgeta**
## What does this PR do?

There are two ways of collecting performance counters in Windows, as explained [here](https://docs.microsoft.com/en-us/windows/win32/perfctrs/collecting-performance-data).

The current Metricbeat implementation uses the one where PdhCollectQueryDataEx is called to retrieve the second raw value that is required to calculate the displayable value.

This PR changes this collection method to a simpler one, the one that calls PdhCollectQueryData twice, with a sleep of one second in between.

The function `randSeq` was implemented in an earlier version in order to support the current implementation, but is not used anymore. The deletion of this function is therefore just considered as cleanup.

## Why is it important?

Collection of the counter "Processor(_Total) % Processor Time" gave false 100 % spikes. Testing on different computers and servers showed that when the sampling frequency, or period, was larger than one second, one would obtain these false 100 % values. They have been checked for being false by comparing with Windows own Performance Monitor. When running several instances of Metricbeat alongside each other, spikes are measured for the instances
- uses the current collection method and
- the period is larger than one second.

This is part of one configuration file used when testing, where the sampling frequency/period is set to 2 seconds:
> metricbeat.modules:
> - module: windows
>   enabled: true
>   metricsets:
>   - perfmon
>   period: 2s
>   perfmon.ignore_non_existent_counters: true
>   perfmon.group_measurements_by_instance: true
>   perfmon.queries:
>   - object: Processor
>     instance:
>     - '_Total'
>     counters:
>     - name: '% Processor Time'

In the two figures below (_Figfure (1)_ and _Figure (2)_), the current implementation (named "thread"), is compared with the implementation of this PR (named "sleep"), for different sampling frequencies/periods. The names "sleep1s" and "thread1s" indicates that the _collection interval_, the time between doing a PdhCollectQueryData call, is hardcoded to 1s in the version.

In _Figure (1)_, each point represents a maximum value over a period of 60 seconds, and explains why the different versions with different frequencies show different values, except for _sleep1s-freq1s_ and _thread1s-freq1s_, which provide almost identical values (as expected). What this figure shows is that the _thread1s-freq2s_ and _thread1s-freq10s_ produce false 100 % spikes.

Zooming in on one of the 100 % spikes has been done in _Figure (2)_, where _thread1s-freq10s_ provides one 100 % value that is most certainly false. In addition, _thread1s-freq2s_ produces 0 % values that does not seem correct either.

_Figure (1)_
thread1s-vs-sleep1s

_Figure (2)_
thread1s-vs-sleep1s-zoom

The figures above show tests performed using Metricbeat 7.17.2, but they have also been reproduced using Metricbeat 8.3.1. They have also been reproduced on different computers and servers, a database server for example. The code in question is the same in both versions, so this is also expected. 

As one also can see from the figures above, and also generally in all the experiments done, the sleep method always produces values that seem correct and that conform with what Windows Performance Monitor is also observing. The problem only arises when the sampling frequency/period is larger than the collection interval (tests have also been made with larger collection intervals, versions of Metricbeat called "sleep2s" and "thread2s", etc.), for the current collection method. It therefore seems like there is some bug concerning the sampling frequency/period of Metricbeat in combination with the creating of the thread. This PR does not explain what this issue is precisely, but simply provides an alternative where this problem does not arise.

In order to get a closer look at what was going on, tests were run where the raw values were fetched after calling the PdhCollectQueryData function. When using either the sleep method, or the thread method with a sampling frequency/period that is not larger than the collection interval (1 second), these raw values were collected at the correct times, and the values could be used to verify the fetched displayable value. Below are some examples of logs of two consecutive samples for these cases. An entry saying "Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5870585175781 133017943902193928 {<nil> 0}}]])", means that the FirstValue of the raw value is 5870585175781 and the SecondValue is 133017943902193928. The displayable value is calculated using taking two consecutive raw values, and then take one minus the difference of the FirstValues divided by the difference of the SecondValues, and at the end multiply by 100 % to get the percent value. The formula is also found [here](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc757283(v=ws.10)). In the below examples, this adds up in all cases.

**sleep1s-freq1s**

> First sample:
> {"@timestamp":"2022-07-09T00:53:10.219+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5870585175781 133017943902193928 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:53:11.223+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 1.5947397042254785 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:53:11.223+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5870595048828 133017943912226976 {<nil> 0}}]])"}
> Consecutive sample:
> {"@timestamp":"2022-07-09T00:53:11.224+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5870595058593 133017943912239452 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:53:12.237+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 4.005576157731239 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:53:12.237+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5870604785156 133017943922371877 {<nil> 0}}]])"}

**sleep1s-freq10s**

> First sample:
> {"@timestamp":"2022-07-08T23:46:47.100+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5831721962890 133017904071006917 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:46:48.116+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 2.633179634240257 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:46:48.116+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5831731855468 133017904081167029 {<nil> 0}}]])"}
> Consecutive sample:
> {"@timestamp":"2022-07-08T23:46:57.092+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5831819648437 133017904170920927 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:46:58.101+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 2.156219690300587 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:46:58.101+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5831829521484 133017904181011550 {<nil> 0}}]])"}

**thread1s-freq1s**

> First sample:
> {"@timestamp":"2022-07-09T00:18:04.609+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5850016337890 133017922846099090 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:18:05.619+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 3.05911239227572 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:18:05.619+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5850026123046 133017922856193031 {<nil> 0}}]])"}
> Consecutive sample:
> {"@timestamp":"2022-07-09T00:18:05.619+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5850026132812 133017922856196130 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:18:06.646+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 3.3303272308510157 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-09T00:18:06.646+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5850036044921 133017922866449717 {<nil> 0}}]])"}

It is, however, not the case when using the current implementation (thread method), with a sampling frequency. Both raw value samples are identical, and hence one does not obtain the fetched displayable value when plugging them into the formula. Here is one example of such logs:

**thread1s-freq10s**

> First sample:
> {"@timestamp":"2022-07-08T23:19:38.104+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5815925039062 133017887781042404 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:19:38.105+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 0.5117438823267872 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:19:38.105+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5815925039062 133017887781042404 {<nil> 0}}]])"}
> Consecutive sample:
> {"@timestamp":"2022-07-08T23:19:48.095+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5816021699218 133017887880951263 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:19:48.097+0200","message":"Formatted values: \\Processor(_Total)\\% Processor Time:[{_Total 5.3058896133528215 {<nil> 0}}]])"}
> {"@timestamp":"2022-07-08T23:19:48.097+0200","message":"Raw values: \\Processor(_Total)\\% Processor Time:[{_Total 5816021699218 133017887880951263 {<nil> 0}}]])"}

Looking into the implementation history, there does not seem to be any explanation of why the current collection method was chosen - especially with this PR's simpler alternative.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Comment by cmacknz** [comment]
The tests won't run until the conflict is resolved it seems.

**Comment by belimawr** [comment]
@jorgeta Thank you so much for this contribution and the super detailed explanation!

I started to look at your code, but the PR is bring many commits that were not made by you, making it pretty hard to review/understand what has changed.

Could you look into it and fix your branch/PR to contain only your commits?

You don't need to worry about keeping your PR up-to-date with `main`, as long as there are no conflicts and the branch was based on a recent version of `main`, it will be fine to merge.

**Review by belimawr — CHANGES_REQUESTED** [review]
Requesting the changes I explained on my previous commit.

**Comment by jorgeta** [comment]
@belimawr Thanks for the reply. The PR should now be updated in the way you requested.

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -71,6 +71,8 @@ https://github.com/elastic/beats/compare/v8.2.0\...main[Check the HEAD diff]
 - Check for nil metadata in GCP. {pull}32281[32281]
 - update kubernetes controllermanager metricset to not collect deprecated metrics and fix dashboard {pull}32037[32037]
 - Fix to ARN parsing for Cloudwatch resource names with leading slashes {pull}32358[32358]
+- update elasticsearch node_stats metricset to use keyword for cgroup memory instead of long {pull}32197[32197]
```
Comment (cmacknz) [inline CHANGELOG.next.asciidoc:74]:
```suggestion
```

This looks like it was added as part of merge conflict resolution and can be removed.

**Comment by cmacknz** [comment]
Some of the lint/check steps are failing for metricbeat:

```
[2022-08-16T17:12:20.283Z] + make -C metricbeat check
[2022-08-16T17:12:20.283Z] make: Entering directory '/var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat'
[2022-08-16T17:12:20.283Z] Installing mage v1.13.0.
[2022-08-16T17:12:20.546Z] /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/.magefile cleaned
[2022-08-16T17:12:20.546Z] mage check
[2022-08-16T17:12:59.336Z] >> fmt - go-licenser: Adding missing headers
[2022-08-16T17:12:59.337Z] >> fmt - goimports: Formatting Go code
[2022-08-16T17:12:59.337Z] >> fmt - autopep8: Formatting Python code
[2022-08-16T17:13:00.766Z] module/windows/perfmon/reader.go
[2022-08-16T17:13:32.895Z] Generated fields.yml for metricbeat to /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat/fields.yml
[2022-08-16T17:13:32.895Z] Generated fields.yml for metricbeat to /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat/fields.yml
[2022-08-16T17:13:32.895Z] >> Building metricbeat.yml for linux/amd64
[2022-08-16T17:13:32.895Z] >> Building metricbeat.reference.yml for linux/amd64
[2022-08-16T17:13:32.895Z] >> Building metricbeat.docker.yml for linux/amd64
[2022-08-16T17:13:32.895Z] Generated fields.yml for metricbeat to /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat/build/fields/fields.all.yml
[2022-08-16T17:15:14.311Z] >> check: Checking source code for common problems
[2022-08-16T17:15:24.335Z] Error: some files are not up-to-date. Run 'make update' then review and commit the changes. Modified: [metricbeat/module/windows/perfmon/reader.go]
[2022-08-16T17:15:24.335Z] ../dev-tools/make/mage.mk:21: recipe for target 'check' failed
[2022-08-16T17:15:24.335Z] make: *** [check] Error 1
[2022-08-16T17:15:24.335Z] make: Leaving directory '/var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat'
script returned exit code 2
```

**Comment by jorgeta** [comment]
> Some of the lint/check steps are failing for metricbeat: https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats/detail/PR-32305/23/pipeline
> 
> ```
> [2022-08-16T17:12:20.283Z] + make -C metricbeat check
> [2022-08-16T17:12:20.283Z] make: Entering directory '/var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat'
> [2022-08-16T17:12:20.283Z] Installing mage v1.13.0.
> [2022-08-16T17:12:20.546Z] /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/.magefile cleaned
> [2022-08-16T17:12:20.546Z] mage check
> [2022-08-16T17:12:59.336Z] >> fmt - go-licenser: Adding missing headers
> [2022-08-16T17:12:59.337Z] >> fmt - goimports: Formatting Go code
> [2022-08-16T17:12:59.337Z] >> fmt - autopep8: Formatting Python code
> [2022-08-16T17:13:00.766Z] module/windows/perfmon/reader.go
> [2022-08-16T17:13:32.895Z] Generated fields.yml for metricbeat to /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat/fields.yml
> [2022-08-16T17:13:32.895Z] Generated fields.yml for metricbeat to /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat/fields.yml
> [2022-08-16T17:13:32.895Z] >> Building metricbeat.yml for linux/amd64
> [2022-08-16T17:13:32.895Z] >> Building metricbeat.reference.yml for linux/amd64
> [2022-08-16T17:13:32.895Z] >> Building metricbeat.docker.yml for linux/amd64
> [2022-08-16T17:13:32.895Z] Generated fields.yml for metricbeat to /var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat/build/fields/fields.all.yml
> [2022-08-16T17:15:14.311Z] >> check: Checking source code for common problems
> [2022-08-16T17:15:24.335Z] Error: some files are not up-to-date. Run 'make update' then review and commit the changes. Modified: [metricbeat/module/windows/perfmon/reader.go]
> [2022-08-16T17:15:24.335Z] ../dev-tools/make/mage.mk:21: recipe for target 'check' failed
> [2022-08-16T17:15:24.335Z] make: *** [check] Error 1
> [2022-08-16T17:15:24.335Z] make: Leaving directory '/var/lib/jenkins/workspace/PR-32305-23-2c910f50-e5a2-4893-9423-401bf7fece6f/src/github.com/elastic/beats/metricbeat'
> script returned exit code 2
> ```

Is this related to the importing of the fmt package? I tried to make some changes unrelated to the change of the collection method, so that the github actions linting would be approved - but maybe I should leave that? Although then the checks would fail in several places.

**Comment by cmacknz** [comment]
It looks like it is just a file that needs to be formatted, if you run `make -C metricbeat check` and then commit the changes it will fix it. This is what it wants changed when I run that command myself:

```diff
diff --git a/metricbeat/module/windows/perfmon/reader.go b/metricbeat/module/windows/perfmon/reader.go
index 04063eb084..645938f882 100644
--- a/metricbeat/module/windows/perfmon/reader.go
+++ b/metricbeat/module/windows/perfmon/reader.go
@@ -24,8 +24,8 @@ import (
 	"fmt"
 	"regexp"
 	"strings"
-	"unicode"
 	"time"
+	"unicode"
 
 	"github.com/elastic/beats/v7/metricbeat/helper/windows/pdh"
 
@@ -132,7 +132,7 @@ func (re *Reader) getValues() (map[string][]pdh.CounterValue, error) {
 	if err != nil {
 		return nil, err
 	}
-	
+
 	// Collect the displayable value.
 	val, err = re.query.GetFormattedCounterValues()
 	if err != nil {
```

File: metricbeat/module/windows/perfmon/reader.go [hunk 2]
```diff
@@ -99,10 +99,10 @@ func (re *Reader) Read() ([]mb.Event, error) {
 	if err := re.query.CollectData(); err != nil {
 		// users can encounter the case no counters are found (services/processes stopped), this should not generate an event with the error message,
 		//could be the case the specific services are started after and picked up by the next RefreshCounterPaths func
-		if err == pdh.PDH_NO_COUNTERS {
+		if err == pdh.PDH_NO_COUNTERS { //nolint:errorlint // Bad linter! This is always errno or nil.
 			re.log.Warnf("%s %v", collectFailedMsg, err)
 		} else {
-			return nil, fmt.Errorf("%v, %w", err, collectFailedMsg)
+			return nil, fmt.Errorf("%v, %w", collectFailedMsg, err)
```
Comment (efd6) [inline metricbeat/module/windows/perfmon/reader.go:105]:
```suggestion
			return nil, fmt.Errorf("%v: %w", collectFailedMsg, err)
```

**Review by efd6 — APPROVED** [review]
Minor nit.

**Comment by narph** [comment]
> Looking into the implementation history, there does not seem to be any explanation of why the current collection method was chosen - especially with this PR's simpler alternative.

@jorgeta, we had this implementation for some time and we switched last year to the `PdhCollectQueryDataEx ` alternative https://github.com/elastic/beats/pull/26886 as it helped with some of the issues we had at that time (missing counter values when configuring a large number of counter queries ).

Have you tested this implementation while configuring a higher number of counters? Are you getting a constant results?

**Comment by jorgeta** [comment]
> > Looking into the implementation history, there does not seem to be any explanation of why the current collection method was chosen - especially with this PR's simpler alternative.
> 
> @jorgeta, we had this implementation for some time and we switched last year to the `PdhCollectQueryDataEx ` alternative #26886 as it helped with some of the issues we had at that time (missing counter values when configuring a large number of counter queries ).
> 
> Have you tested this implementation while configuring a higher number of counters? Are you getting a constant results?

Hi, @narph - thanks for your feedback. We have tested the implementation for a different amount of counters; from having just the one, and up to 90 "windows.perfmon.*" counters, also along with additional 54 "system.*" (that is, tests have collected between 1 and 144 counters in total). I could try to test with a larger number - if this is not considered large.

So, as far as I can see, the implementation in this PR is still different from the one that was before `PdhCollectQueryDataEx` was added. Seems like in the earlier implementation, the `PhdCollectQueryData` was only called once per collection. This means that if one is collecting counters every 15 seconds, then `PdhGetFormattedCounterValueDouble` would use the raw values collected 15 seconds ago along with the ones that were just collected, in order to compute the displayable value. When `PhdCollectQueryDataEx` was added, this problem was solved. This PR only wants to call `PhdCollectQueryData` twice with one second of sleep in between for every time a (rate) counter is collected, contrary to what I can see must have been happening in the pre-`PhdCollectQueryDataEx` implementation.

**Review by belimawr — CHANGES_REQUESTED** [review]
I'm no Windows expert, but it looks good to me. Just the small changelog fix that's needed.

**Comment by cmacknz** [comment]
@fearful-symmetry can you review this as well?

@jorgeta sorry this is taking so long, this is a tricky piece of code with a history of issues as I'm sure you know.

**Review by belimawr — APPROVED** [review]
All good from my side, but, please, wait for @fearful-symmetry or someone with more expertise on this area than me.

**Review by fearful-symmetry — APPROVED** [review]
So, I don't have much windows knowledge, but just based on the bits I do understand, this LGTM.

**Comment by cmacknz** [comment]
Thanks everyone, merging this with 8.5 as the target version. We can have our QA team check this on a few different versions of Windows as a final sanity check.

**Comment by cmacknz** [comment]
Sure, this is effectively a bug fix so we can backport to 7.17

**Pull request merged**